Add a threshold when comparing screen order for selectables.

hannah-hyj opened the pull request:
Add a threshold when comparing screen order for selectables. So when the vertical position diff is within the threshold, will compare the horizontal position.

This fixes https://github.com/flutter/flutter/issues/111021 and https://github.com/flutter/flutter/issues/127942

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

hannah-hyj committed:
Update selectable_region.dart

hannah-hyj committed:
Update selectable_region.dart

hannah-hyj committed:
Update selectable_region_test.dart

hannah-hyj committed:
Update selectable_region_test.dart

hannah-hyj committed:
Update selectable_region_test.dart

hannah-hyj committed:
Update selectable_region.dart

chunhtai commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -39,6 +39,9 @@ const Set<PointerDeviceKind> _kLongPressSelectionDevices = <PointerDeviceKind>{
 };
 
+
+const double _kSelectableComparingThreshold = 5.0;
```
This would likely break if there is more padding in widget span?

chunhtai replied:
Also this should probably call vertical threshold?

hangyujin replied:
renamed to _kSelectableVerticalComparingThreshold

chunhtai reviewed (commented):
This solution seems a bit fragile. We may soon run into situation where people would require conflicting thresholds for different use cases.

hannah-hyj committed:
Update selectable_region.dart

Update selectable_region.dart

Update selectable_region.dart

Update selectable_region_test.dart

test

hangyujin commented:
Yes, this fix is not a one-and-done solution, but I don't think this fix will make things worse.

in #111021 and #127942, both cases  the widgetspan just shifted like 0.5 pixel. This fix can resolve some existing issues.

A final solution will probably be exposing compareOrder function to developers and we can follow up if there are more requests.

Renzo-Olivares reviewed (commented):
I think this change is reasonable if it doesn't break any current screen order calculations. I've experienced this problem where the rects are off by a few pixels so they do not defer to horizontal comparison.

chunhtai commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -39,6 +39,9 @@ const Set<PointerDeviceKind> _kLongPressSelectionDevices = <PointerDeviceKind>{
 };
 
+
+const double _kSelectableVerticalComparingThreshold = 3.0;
```
Can you add a comment explaining the situation why this is needed

chunhtai reviewed (approved):
LGTM
